usr/share/rear/backup/BORG/default/500_make_backup.sh

```diff
@@ -18,20 +18,36 @@ done
 
 # User might specify some additional output options in Borg.
 # Output shown by Borg is not controlled by `rear --verbose' nor `rear --debug'
+# only, if BORGBACKUP_SHOW_PROGRESS is true.
 local borg_additional_options=''
 
 is_true $BORGBACKUP_SHOW_PROGRESS && borg_additional_options+='--progress '
 is_true $BORGBACKUP_SHOW_STATS && borg_additional_options+='--stats '
+is_true $BORGBACKUP_SHOW_LIST && borg_additional_options+='--list --filter=AME '
+is_true $BORGBACKUP_SHOW_RC && borg_additional_options+='--show-rc '
+is_true $BORGBACKUP_EXCLUDE_CACHES && borg_additional_options+='--exclude-caches '
+is_true $BORGBACKUP_EXCLUDE_IF_NOBACKUP && borg_additional_options+='--exclude-if-present .nobackup '
+[[ -n $BORGBACKUP_TIMESTAMP ]] && borg_additional_options+="--timestamp $BORGBACKUP_TIMESTAMP "
+
+# Borg writes all log output to stderr by default.
+# See https://borgbackup.readthedocs.io/en/stable/usage/general.html#logging
+#
+# If we want to have log output from Borg appearing in rear logs, we don't have
+# to do anything, since Borg logs to stderr and that is what rear is saving in
+# Logfile.
+#
+# If `--progress` is used for `borg create` we don't want the output in rear
+# log file, since it contains control sequences. If not used, we want Borg
+# output in rear log file, the amount of logs written by Borg is determined by
+# other options above e.g. by `--stats` or `--list --filter=AME`.
 
 # Start actual Borg backup.
-Log "Creating archive ${BORGBACKUP_ARCHIVE_PREFIX}_$BORGBACKUP_SUFFIX \
-in repository $BORGBACKUP_REPO"
-
-borg create --one-file-system $borg_additional_options $verbose \
-$BORGBACKUP_OPT_COMPRESSION $BORGBACKUP_OPT_REMOTE_PATH \
-$BORGBACKUP_OPT_UMASK --exclude-from $TMP_DIR/backup-exclude.txt \
-${borg_dst_dev}${BORGBACKUP_REPO}::\
-${BORGBACKUP_ARCHIVE_PREFIX}_$BORGBACKUP_SUFFIX \
-${include_list[@]} 0<&6 1>&7 2>&8
+if is_true $BORGBACKUP_SHOW_PROGRESS; then
+  borg_create 0<&6 1>&7 2>&8
+elif is_true $VERBOSE; then
+  borg_create 0<&6 1>&7 2> >(tee >(cat 1>&2) >&8)
```

Comment:
@flyinggreenfrog 
could you explain what your intent behind the
`0<&6 1>&7 2> >(tee >(cat 1>&2) >&8)`
stderr redirection plus process substitution is?

If what you like to get when $VERBOSE is set
that the borg_create stderr output appears on the user's terminal
and additionally in rear's log file I think the following code
could be simpler and more straightforward and easier to see what the intent is
because the variable name RUNTIME_LOGFILE indicates the intent behind
(but I did not yet test how that actually behaves "inside" rear):
```
borg_create 0<&6 1>&7 2> >( tee -a $RUNTIME_LOGFILE )
```
which lets the `tee` stdout output appear where stdout is while `tee` runs
which could be still the default which is rear's log file
(depending on when `1>&7` is evaluated by bash)
so you may have to use something like
```
borg_create 0<&6 1>&7 2> >( tee -a $RUNTIME_LOGFILE 1>&8 )
```
to ensure the `tee` stdout output appears where stderr is set to
on the user's terminal when he launched rear.
I hope the fd8 file descriptor `&8` of the rear parent process
is inherited to the `tee` process substitution.

Reply:
Maybe as general overview, what I have in mind.

With the current implementation, we have roughly 3 points, where I want to see Borg log output going to the rear logfile:

- `borg create` and `borg extract`: `0<&6 1>&7 2>&8` is used at the moment (no Borg stderr output is going to logfile)
- `borg prune`: no redirection is used at the moment (all Borg stderr output is going to logfile)

Generally I understand https://github.com/rear/rear/wiki/Coding-Style#what-to-do-with-stdin-stdout-and-stderr, that in case interaction with the user is needed (either while making backup or while restore), redirection has to take place.

In how I'm using Borg (with `--list`, `--stats`, `--show-rc`, where applicable), it would be nice to have such output in the rear logfile. That is normally I wouldn't need any redirection.

Then depending on VERBOSE (or `-v|--verbose`) additionally make it visible to user, that he sees, what's going on, e.g. via restore to see the list of files restored.

So for me `0<&6 1>&7` for `borg create` is not necessary, or? Why was it implement like that? And `borg prune` had no redirection.

Only problem is in case you use `--progress` for borg, as then some control characters are used for the progress output, which kind of would mess up the logfile. Hence the 3 cases (progress, verbose and other).

For `borg extract` it makes more sense for me to set `0<&6 1>&7 2>&8`, since normally you have to interactively select archive to restore from.

But here also logging of `borg extract` output to logfile in case no `--progress` is used, would be good.

So my general question is, which redirections are really needed?

I'm not so into the rear details, but if you think working with RUNTIME_LOGFILE is preferred, I will test that.

Reply:
I am not a Borg user so I can neither test anything with Borg
nor can I tell what Borg program output could be needed
to be shown on the user's terminal.

Using `tee -a $RUNTIME_LOGFILE` would be my personal
preferred method because personally for me it is very important
that others can easily understand what the ReaR scripts do
because ReaR is meant that the users can and should adapt
or extend the ReaR scripts as needed to make things work
for their specific use cases.

For me both `0<&6 1>&7 2> >( tee -a $RUNTIME_LOGFILE )`
and `0<&6 1>&7 2> >( tee -a $RUNTIME_LOGFILE 1>&8 )`
work ok but the latter is better because it still works
correctly when the user redirects his stderr e.g. like
```
# rear mkrescue 2>/tmp/rearstderr
```

Reply:
Attention:
I think when using process substitution we must ensure
to append to the filename that is the result of the process substitution
because when we do not append we overwrite existing things in that file.
So I think we must use `2>> >( COMMAND )` instead of `2> >( COMMAND )`
for example
```
0<&6 1>&7 2>> >( tee >( cat 1>&2 ) 1>&8 )
0<&6 1>&7 2>> >( tee -a $RUNTIME_LOGFILE 1>&8 )
```
or for both stdout and stderr
```
0<&6 1>> >( tee >( cat 1>&2 ) 1>&7 ) 2>> >( tee >( cat 1>&2 ) 1>&8 )
0<&6 1>> >( tee -a $RUNTIME_LOGFILE 1>&7 ) 2>> >( tee -a $RUNTIME_LOGFILE 1>&8 )
```

Reply:
See e57f8eb465a0bac73bf7edf20b0b8a4864f1e6f9, 7e966418d15d6f2893579d2b0db7c0494cf65f88 and 66ead8ad41485f44f8a7a16cdcaadf1e7281c3c8.

Reply:
@flyinggreenfrog
thank you for your recent commits that use `tee -a $RUNTIME_LOGFILE`
that ensures writing to the log file happens in append mode, cf.
https://unix.stackexchange.com/questions/346062/concurrent-writing-to-a-log-file-from-many-processes

Meanwhile I think when using process substitution `>( COMMAND )`
we could use both `2>> >( COMMAND )` and `2> >( COMMAND )`
because `>( COMMAND )` evaluates to a file name where to the
current bash writes and where from COMMAND gets its stdin.
Usually that intermediate file is a named pipes (FIFO).
I think it does not matter if the current bash writes or appends to
that intermediate file because the current bash is a single process
so that both methods result that all is written sequentially and
nothing is overwritten by another process (i.e. the current bash is
the only writer to that intermediate file).
But what matters is that COMMAND appends to the ReaR log file
because when using process substitution COMMAND is run
asynchronously as another process beside the current bash.
Because the current bash has stdout redirected to the log file
now the process substitution `2> >( COMMAND )` also writes
asynchronously to the ReaR log file at the same time while
the current bash stdout appends to the ReaR log file.

With COMMAND = `tee -a $RUNTIME_LOGFILE`
I am sure it appends to the ReaR log file.

In contrast with COMMAND = `tee >( cat 1>&2 )`
I am not sure if that actually appends to the ReaR log file.
I think it could actually append to the ReaR log file if the
result of the inner process substitution ` cat 1>&2` is
that the `cat` stdout file descriptor is a copy of the
current bash's stderr file descriptor which is
redirected to append to the log file so that
also `cat` actually appends to the log file.
I think the `tee >( ... )` plain stdout writing (no append)
does not matter as described above because `tee` is a single
process that writes into a FIFO wherefrom `cat` gets its stdin.
What matters is whether or not `cat` stdout plain writes
or actually appends to the ReaR log file.

What I do not know is what excatly it means when a file descriptor
is duplicated as in `1>&2` i.e. what excatly `dup` does.
Usual documentation often talks about 'copy' of file descriptors
but what I would like to know is what in the kernel's file access
structures gets copied so that it exists as two separated things
in particular when the current file position / file offset value gets
copied when several process are accessing one same file, cf.
https://stackoverflow.com/questions/5256599/what-are-file-descriptors-explained-in-simple-terms

Reply:
Thanks for your explanations.

Is there anything that I still should change in the current PR? Or you just wanted to add some background explanation to this discussion? @jsmeix 

For me with the usage of `tee -a $RUNTIME_LOGFILE` things look to be ok.

Reply:
I only liked to add some background explanation to this discussion.

The reason behind is that I intend to describe in
https://github.com/rear/rear/wiki/Coding-Style
a properly working method how to show stdout and stderr
both on the user's terminal and also in ReaR's log file.